### PR TITLE
3D-Improved-UNet pull request Yitian Guo s4706162

### PR DESCRIPTION
**Pull request by OnenightinBJ**
### **Overview**: This Pull Request implements the 3D Improved UNet model for segmenting prostate MRI image data, and provides comprehensive improvements and documentation for model training, validation, and testing.
**Main Changes:**

- **Model implementation**: an implementation of the 3D Improved UNet model has been added to the modules.py, including residual linking and attention mechanisms to improve model training efficiency and segmentation accuracy.

- **Data loader**: A data loading and preprocessing module has been implemented in the dataset.py file to support the processing of 3D prostate MRI datasets.

- **Training script**: Logic for training, validating and testing the model is added in the train.py file, and losses and metrics during training are plotted.

- **Prediction Scripts**: Examples of model predictions are provided in the predict.py file, including model inference and visualization of results.

- **Documentation update**: The README.md file has been updated to describe in detail the project's goals, model architecture, usage, and dependencies.

**Purpose**: These changes are intended to enable a reliable automated segmentation tool to aid in the processing of 3D prostate MRI data and to achieve dice coefficients above 0.7 in each category.

### CODE REVIEW

**Comment by celinsze** [comment]
**This is an initial inspection, no action is required at this point**

Recognition problem: 
Problem Solving: 
•	Algorithm solves the problem and achieves a Dice score of above 0.8 for all labels. 
Algorithm implementation: 
•	Algorithm implemented in PyTorch for the UNet3D model and functions.
•	Test set is split in predict.py for model testing instead of using the same split used in train.py. Same settings used for data splitting from initial inspection but add risk for overlapping between train and test set.
Design: 
•	Modularised and reusable code.
Commenting: 
•	Some comments throughout the code. 
Difficulty: 
•	3D UNet with residual blocks and attention gate included for each layer.

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files (gitignore). Please update for merge, doesn't affect grade.
  
Segmentations not one hot encoded -2 additionally

**Comment by OnenightinBJ** [comment]
Hi @shakes76,
I have resolved the conflict in `.gitignore` and updated the branch. Please review and merge when convenient. Thank you!
